File: src/algorithms/sets/disjoint-sets/union-find/sources/union-find.py
# Union-Find (Disjoint Set Union) — Path Compression + Union by Rank
# Maintains a partition of elements into disjoint sets.
# find(x): returns the root representative of x's set, compressing the path.
# union(x, y): merges the sets containing x and y using rank heuristic.
# Time: O(α(n)) amortized per operation — Space: O(n)


def union_find(element_count: int, operations: list[list[int]]) -> dict:
    parent = list(range(element_count))  # @step:initialize
    rank = [0] * element_count  # @step:initialize

    def find(element: int) -> int:  # @step:find-root
        if parent[element] != element:
            parent[element] = find(parent[element])  # @step:find-root
        return parent[element]

    def union(elem_a: int, elem_b: int) -> None:
        root_a = find(elem_a)  # @step:find-root
        root_b = find(elem_b)  # @step:find-root
        if root_a == root_b:
            return

        if rank[root_a] >= rank[root_b]:
            parent[root_b] = root_a  # @step:union-sets
            if rank[root_a] == rank[root_b]:
                rank[root_a] += 1
        else:
            parent[root_a] = root_b  # @step:union-sets

    for elem_a, elem_b in operations:
        union(elem_a, elem_b)

    # Build final components
    component_map: dict[int, list[int]] = {}
    for elem_idx in range(element_count):
        root = find(elem_idx)
        if root not in component_map:
            component_map[root] = []
        component_map[root].append(elem_idx)

    return {"components": list(component_map.values())}  # @step:complete

File: src/algorithms/sets/disjoint-sets/union-find/sources/union-find.go
// Union-Find (Disjoint Set Union) — Path Compression + Union by Rank
// Maintains a partition of elements into disjoint sets.
// find(x): returns the root representative of x's set, compressing the path.
// union(x, y): merges the sets containing x and y using rank heuristic.
// Time: O(α(n)) amortized per operation — Space: O(n)

package main

import "fmt"

func findRoot(parent []int, element int) int {
	// @step:find-root
	if parent[element] != element {
		parent[element] = findRoot(parent, parent[element]) // @step:find-root
	}
	return parent[element]
}

func unionSets(parent []int, rank []int, elemA int, elemB int) {
	rootA := findRoot(parent, elemA) // @step:find-root
	rootB := findRoot(parent, elemB) // @step:find-root
	if rootA == rootB {
		return
	}

	if rank[rootA] >= rank[rootB] {
		parent[rootB] = rootA // @step:union-sets
		if rank[rootA] == rank[rootB] {
			rank[rootA]++
		}
	} else {
		parent[rootA] = rootB // @step:union-sets
	}
}

type Operation struct {
	elemA, elemB int
}

func unionFind(elementCount int, operations []Operation) [][]int {
	parent := make([]int, elementCount) // @step:initialize
	rank := make([]int, elementCount)   // @step:initialize
	for idx := range parent {
		parent[idx] = idx
	}

	for _, op := range operations {
		unionSets(parent, rank, op.elemA, op.elemB)
	}

	// Build final components
	componentMap := make(map[int][]int)
	for elemIdx := 0; elemIdx < elementCount; elemIdx++ {
		root := findRoot(parent, elemIdx)
		componentMap[root] = append(componentMap[root], elemIdx)
	}

	components := make([][]int, 0, len(componentMap))
	for _, group := range componentMap {
		components = append(components, group)
	}
	return components // @step:complete
}

func main() {
	operations := []Operation{{0, 1}, {2, 3}, {4, 5}, {6, 7}, {0, 2}, {4, 6}, {0, 4}}
	components := unionFind(8, operations)
	fmt.Println(components)
}
